### PR TITLE
Doxygen and swig

### LINKED ISSUE #249: pass doxygen to Python
2 options:

- doxyg2swig, at least one version is at https://github.com/m7thon/doxy2swig
- Next version of SWIG will do this natively: Doxygen support was merged into swig master on 2018-06-07. See https://github.com/swig/swig/blob/master/Doc/Manual/Doxygen.html

### PR DESCRIPTION
**Pull request by KrisThielemans**
Use doxygen and doxy2swig to add docstrings to swig
    
- Run doxygen automatically (don't need to use "make doc" anymore)
- Use doxy2swig to pass doxygen comments to SWIG-Python etc.
- Include doxy2swig from https://github.com/m7thon/doxy2swig
    
Although new swig will natively support doxygen, it's probably better to
use doxy2swig for the moment.
    
Closes #249

### CODE REVIEW

**Comment by ashgillman** [comment]
What's the timeline for this? Is this for before the course?

If there is the option of using Doxygen rather than doxy2swig it may be preferable. It looks like doxy2swig hasn't had any commits over the last 18 months.

Does it run yet? I can check it compiles with my NixBuild.

**Comment by KrisThielemans** [comment]
sure it runs! :-)
Needs `BUILD_DOCUMENTAION=ON`

Works nicely for me. It'd be nice to have it for the course, as this allows nice things like 
```python
help(stir.Scanner)
```

**Comment by KrisThielemans** [comment]
Not sure when next SWIG release will be out, but we can't wait for that... Hardly anyone will have it.

**Comment by rijobro** [comment]
Works great for me (only tested `help(stir.Scanner)`. Two small things:

1. `make -j4` fails. I have to use `make` until the `Doxy2swig` stuff is finished, then I can cancel and go back to `make -j4`.
2. `Doxy2swig` produces a lot of output to the terminal. Could potentially be directed to a log file instead. (This one clearly isn't very important.)

**Comment by KrisThielemans** [comment]
I'm afraid I have to accept this. Will need to sort out issues later on.

**Comment by KrisThielemans** [comment]
> make -j4 fails. I have to use make until the Doxy2swig stuff is finished, then I can cancel and go back to make -j4.

ah. just saw this. I thought i had put in all the relevant dependencies to prevent this problem. oh well.

File: src/CMakeLists.txt [hunk 1]
```diff
@@ -255,13 +251,31 @@ if(BUILD_DOCUMENTATION)
...
-    add_custom_target(doc
+    # The following lines are modifications of Jannis Fischer's original. The modifications were inspired by
+    # https://samthursfield.wordpress.com/2015/11/21/cmake-dependencies-between-targets-and-files-and-custom-commands/
+
+    # Add a custom command that does everything, but also generates a doxygen.stamp file.
+    # This should prevent doxygen being re-run when not needed.
+    # TODO This should depnd on all .h files really
+    add_custom_command(
+        OUTPUT doxygen.stamp
+        DEPENDS ${doxyfile}
```
[the remark this answers is not in the document]
Comment (NikEfth) [reply src/CMakeLists.txt]: It was for qtcreator, but I think I never pushed it.

**Comment by NikEfth** [comment]
I think the required version of SWIG should better change as: 
> To attach docstrings to classes with SWIG for python and using the -builtin option, a version of SWIG >= 3.0.7 is required. Without this, the class documentation including constructor lists (-c) and attribute lists (-a) will not be available from python.

**Comment by NikEfth** [comment]
> make -j4 fails. I have to use make until the Doxy2swig stuff is finished, then I can cancel and go back to make -j4.
Cannot confirm this. I just run everything with `make -j5`

**Comment by NikEfth** [comment]
I found reports that `doxy2swig` works only with python2. We could check ` PYTHON_VERSION_STRING `to see the version of `PYTHON_EXECUTABLE` before running.

File: src/CMakeLists.txt (hunk 1 above)
[the remark this answers is not in the document]
Comment (NikEfth) [reply src/CMakeLists.txt]: Found it:
https://github.com/UCL/STIR/blob/d7881273eeda35388dc68b2acf41e9df7f3e1f48/src/cmake/stir_exe_targets.cmake#L30

**Comment by rijobro** [comment]
> I found reports that doxy2swig works only with python2

I saw this, too, but it worked fine for me with `python3` (with -j1).

File: src/swig/CMakeLists.txt [hunk 2]
```diff
@@ -28,6 +28,35 @@ if(BUILD_SWIG_PYTHON OR BUILD_SWIG_OCTAVE OR BUILD_SWIG_MATLAB)
...
+    )
+    add_custom_target(
+      doxy2swig
+      DEPENDS STIR_DOXY2SWIG.i)
+    # Set a variable that will be passed to SWIG. It specifies the location of the generated .i file.
+    # We need some interesting "escaping" to be able to pass a quoted string as a preprocessor variable to SWIG.
+    # (The value of the preprocessor variables needs to have the quotes, as SWIG doesn't
+    # substitute preprocessor variables inside quotes).
+    # (SWIG_DOXY_OPTIONS will be empty if BUILD_DOCUMENTATION is off).
+    set(SWIG_DOXY_OPTIONS -DDOXY2SWIG_XML_INCLUDE_FILE="\\\"${CMAKE_CURRENT_BINARY_DIR}/STIR_DOXY2SWIG.i\\\"")
```
[the remark this answers is not in the document]
Comment (KrisThielemans) [reply src/swig/CMakeLists.txt]: it works fine actually

**Comment by KrisThielemans** [comment]
> make -j4 fails.
@rijobro could you confirm which CMake and make version you're using?

**Comment by KrisThielemans** [comment]
> Doxy2swig produces a lot of output to the terminal. Could potentially be directed to a log file instead.

I'm not going to do this. doxygen produces a ton as well, but so does the compilation. It would be non-standard to redirect to a file automatically as then any warnings might go unnoticed.

**Comment by KrisThielemans** [comment]
@NikEfth @ashgillman @rijobro could you give this another spin? I've cleaned this up now and it should be ready. The remaining worry is the parallel build, but I've investigated this more and can't see anything wrong with the generated makefiles. Also no immediate warnings bells on the CMake list. See [here](https://www.mail-archive.com/cmake@cmake.org/msg59978.html).

To test the parallel build, I think we have to do the following to test the dependencies as much as possible. (starting from a new one)
```
# first build without doc (and without executables to save some time)
cmake -DBUILD_EXECUTABLES=OFF -DBUILD_DOCUMENTATION=OFF -DBUILD_SWIG_PYTHON=OFF <source>
make -j4
# now reconfigure
cmake -DBUILD_DOCUMENTATION=ON -DBUILD_SWIG_PYTHON=ON .
make -j4
```
use as high a number for `-j`- as possible for you of course.

**Comment by rijobro** [comment]
Cmake version 3.11.1.

**Comment by KrisThielemans** [comment]
`make`?

**Comment by rijobro** [comment]
Just reran since all your new commits, and `make -j4` now works fine for me.

(make version is `GNU Make 3.81`)

**Pull request merged**